src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java

```diff
@@ -644,8 +644,594 @@ public void invalidMultiplyAndAssignmentOperationExpr_voidType() {
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.BOOLEAN_OBJECT, "y");
+    AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr);
+    // No exception thrown, so we succeeded.
+  }
+
+  @Test
+  public void validXorAssignmentOperationExpr_booleanBoxedTypeWithUnboxedType() {
+    // Swap case in "validXorAssignmentOperationExpr_booleanWithBooleanBoxedType".
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN_OBJECT, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.BOOLEAN, "y");
+    AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr);
+    // No exception thrown, so we succeeded.
+  }
+
+  @Test
+  public void validXorAssignmentOperationExpr_booleanBoxedType() {
+    // No need swap case.
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN_OBJECT, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.BOOLEAN_OBJECT, "y");
+    AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr);
+    // No exception thrown, so we succeeded.
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_booleanWithNumericTypes() {
+    // Swap case in "invalidORAssignmentOperationExpr_integerWithBooleanTypes".
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.INT, "y");
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_booleanWithIntegerBoxedTypes() {
+    // Swap case in "invalidORAssignmentOperationExpr_integerBoxedTypeWithBooleanTypes".
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.INT_OBJECT, "y");
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_booleanWithReferencedTypes() {
+    // Swap case in "invalidORAssignmentOperationExpr_referencedTypeWithBooleanTypes".
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.STRING, "y");
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_booleanWithNullTypes() {
+    // No valid swap case.
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN, "x");
+    ValueExpr rhsExpr = ValueExpr.withValue(NullObjectValue.create());
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_booleanWithNewObjectTypes() {
```

Comment:
Nit: `invalidXor`

Reply:
oops. Copy and paste a lot. Fixed right now. Should be good.